9-writing-advanced-plans/README.md

```diff
@@ -11,258 +11,249 @@ In this exercise you will further explore Puppet Plans:
-
-print(json.dumps({'answer': bool(random.getrandbits(1))}))
-```
-
-Then we can write out the plan. Save the following as `modules/exercise9/plans/yesorno.pp`:
-
-```puppet
-plan exercise9::yesorno (TargetSpec $nodes) {
-  $results = run_task('exercise9::yesorno', $nodes)
-  $subset = $results.filter |$result| { $result[answer] == true }.map |$result| { $result.target }
-  run_command("uptime", $subset)
-}
-```
-
-In the above plan we:
-
-* Accept a comma-separated list of nodes
-* Run the `exercise9::yesorno` task from above on all of our nodes
-* Store the results of running the task in the variable `$results`. This will contain a `ResultSet` containing a list of `Result` objects for each node and the data parsed from the JSON response from the task
-* We filter the list of results to get the node names for only those that answered `true`, stored in the `$subset` variable
-* We finally run the `uptime` command on our filtered list of nodes
-
-You can see this plan in action by running:
-
-```bash
-bolt plan run exercise9::yesorno nodes=all --modulepath ./modules
-```
-
-When run you should see output like the following. Running it multiple times should result in different output, as the return value of the task is random the command should run on a different subset of nodes each time.
-
-```bash
-[
-  {
-    "node": "node1",
-    "status": "success",
-    "result": {
-      "stdout": " 23:41:49 up 8 min,  0 users,  load average: 0.00, 0.03, 0.04\n",
-      "stderr": "",
-      "exit_code": 0
-    }
-  },
-  {
-    "node": "node2",
-    "status": "success",
-    "result": {
-      "stdout": " 23:41:49 up 7 min,  0 users,  load average: 0.32, 0.08, 0.05\n",
-      "stderr": "",
-      "exit_code": 0
+In the previous exercise you ran tasks and commands within the context of a plan. Now you will create a task that captures the return values and uses those values in subsequent steps. The ability to use the output of a task as the input to another task allows for creating much more complex and powerful plans. Real-world uses for this might include:
+
+* A plan that uses a task to check how long since a machine was last rebooted, and then runs another task to reboot the machine on nodes that have been up for more than a week.
+* A plan that uses a task to identify the operating system of a machine and then run a different task on each different operating system.
+
+1. Create a task that will print a JSON structure with an `answer` key with a value of true or false. The important thing to note is the use of JSON to structure our return value. 
+
+    ```python
+    #!/usr/bin/env python
+    
+    """
+    This script returns a JSON string with a single key, answer which
+    has a boolean value. It should flip between returning true and false
+    at random
+    """
+    
+    import json
+    import random
+    
+    print(json.dumps({'answer': bool(random.getrandbits(1))}))
+    ```
+2. Save the task as `modules/exercise9/tasks/yesorno.py`.
```

Comment:
This should be part of step one. Most users will want to specify the file name before they start editing.

Reply:
Fixed